crates/apollo-compiler/src/parser.rs

```diff
@@ -362,12 +362,23 @@ impl SourceFile {
         })
     }
 
-    pub fn get_line_column(&self, index: usize) -> Option<(usize, usize)> {
+    pub(crate) fn get_line_column(&self, index: usize) -> Option<LineColumn> {
         let (_, line, column) = self.ariadne().get_byte_line(index)?;
-        Some((line, column))
+        Some(LineColumn { line, column })
     }
 }
 
+/// A specific location within the source of a file, like an offset.
+///
+/// See also [`crate::node::LineColumnRange`].
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub struct LineColumn {
+    /// The line number, starting at 0 for the first line (not 1 like [`GraphQLError`]).
+    pub line: usize,
+    /// The column number, starting at 0 for the first character of the line.
+    pub column: usize,
+}
```

Comment:
It seems awkward to have two different structs in the public API that are the same except off by one. @goto-bus-stop what do you think?

Reply:
Yeah, the challenge is to draw a distinction between a generic location in source code and the definition of location within a GraphQL Error at runtime. They serve quite different purposes, in practice, but they sure look similar 🤔

Reply:
I’m questioning whether the distinction needs to exist at all in this library’s public API. Arguably 1-indexing is "more generic" since that’s how humans usually talk about line numbers. So apollo-compiler could have one line-column struct with a generic name that is also used in GraphQL errors, and conversion to 0-indexed for LSP purpose could happen in the language server. (The redundant +1 then -1 computation is extremely cheap.)

Reply:
Cheap, yes, but very easy to make mistakes on, since we need to do them in several different repos today. I have already fixed several bugs relating to a + or -1 being missing in one line or another 😬. Feels bad to do it in more places than necessary.

If we have two canonical types, the + and -1 only need to be in a single place, and we can reference the correct type for each use-case from the rest.

Reply:
If we have both in the same library, especially because this is error prone, I think they should not be called "generic" and "GraphQL" but something like `ZeroIndexedLineColumn` and `OneIndexedLineColumn`. Kinda unpleasant to use but makes the difference very obvious.

That said, couldn’t the language server define an `LspLineColumn` type in one place and only do +1 or -1 in `From` conversions to or from that type?

Reply:
I guess it is a bit annoying for lsps because you are probably dealing with `lsp_types::Position` and `apollo_compiler::LineColumn`, and you can't implement `From<>` on types from external crates. You'd probably use a helper function instead of a `From<>` impl.
But I'd also prefer having just one type for line/column in apollo-compiler... We already have multiple location types (indices and line/column)

Reply:
I would rather that the compiler defines the line/column range that the lsp would use. I don't want them to deal with incrementing/decrementing it themselves, especially since it sounds like we've had bugs with several implementations already. Does it need to be different structs though? If `LineColumn` is used in context of a `Diagnostic` or `GraphQLError`, it's has an all together different meaning then when used in a context of a `Node`. Or is the worry that when looking at a struct definition of `Node` and `Diagnostic` and see `LineColumn` the two are actually not identical in range?

Reply:
For me, the main benefit of two structs is using type checking to enforce offset checking. Meaning, if you do have a function that converts from `LineColumn` to `LspLineColumn` (or whatever), that function should _always_ work for those types. If the info about whether it's zero- or one-indexed is not in the struct itself (but instead some parent struct like `Diagnostic`), then you wouldn't be able to lean on the compiler to ensure that conversion is right.

Worth pointing out as well, I think, that there are _many_ location types spread around our various repos right now. I'd love to start consolidating those however possible 😁. There are 3 similar types just in `federation-rs`! I'm not sure if we want all of those crates to depend on this one, though.

Reply:
That is assuming that only the LSP needs 0-indexed positions. I'm not sure that's true. It also _feels_ bad for this crate to be constantly converting _to_ 1-indexed for every location just for the consumer (LSP) to be subtracting again. I guess in practice it's a very quick operation, but my brain doesn't like it 😆.

Reply:
Yeah. I think it's preferable to have just one. An LSP needs 0-indexed but other use cases (including graphql server error responses) need 1-indexed. Someone always has to convert, and IMO conversion is preferable to having two APIs that return different indexes (even if it's a different type). apollo-rs provides graphql JSON error types so we have a requirement in the crate to *at least* have a 1-indexed type. I think we need a *strong* reason to add another one...

Reply:
Since the main hold-up is about the second struct, I removed it and switched everything back to 1-indexed. So now the only change in this PR is returning `Range` from a few places instead of single locations.

I'll move the zero-indexed struct to a different shared crate (`apollo-composition`, probably) so that everyone that needs to convert can re-use that one.